### PR TITLE
Optimize and make visitor generic

### PR DESCRIPTION
**Pull request by turbolent**
## Description

- Make the AST visitors generic
- Refactor the `Accept` method on `Element` to a global function, as Go does not support generic receiver methods (ideally we would like to have `Element` have a function `Accept[T](visitor Visitor[T]) T`)

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -637,7 +637,7 @@ func (interpreter *Interpreter) RecoverErrors(onError func(error)) {
 	}
 }
 
-func (interpreter *Interpreter) VisitProgram(program *ast.Program) ast.Repr {
+func (interpreter *Interpreter) VisitProgram(program *ast.Program) any {
```
Comment (dsainati1) [inline runtime/interpreter/interpreter.go:640]: Why are we using `any` here? Can we have these interpreter functions return `Value` like the checker ones return `Type`?

**Review by dsainati1 — COMMENTED** [review]
This is awesome. Very excited for this!

File: runtime/interpreter/interpreter.go (hunk 1 above)
Comment (turbolent) [reply runtime/interpreter/interpreter.go:640]: Unfortunately there are currently two types being returned, `Value` and `controlReturn`. 

That is why for now I kept the interpreter code as-is (`ast.Repr` was an alias for `any`). 
In a subsequent PR I was planning on refactoring this to a single "enum" `Result`, which can represent both.

**Review by dsainati1 — APPROVED** [review]

File: runtime/interpreter/interpreter.go (hunk 1 above)
Comment (turbolent) [reply runtime/interpreter/interpreter.go:640]: Continued the improvements in #1919

**Review by dreamsmasher — APPROVED** [review]
LGTM, splitting the Visitor traits by target AST-node type is a nice touch.

**Pull request merged**